### PR TITLE
gnupg: add a patch to fix emacs easypg

### PR DESCRIPTION
**Pull request by kurnevsky**
###### Description of changes

gnupg 2.4.1 update breaks emacs easypg, see https://dev.gnupg.org/T6481

cc @fpletz @vrthra

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kurnevsky** [comment]
Temporarily changed target to run nix-review.

**Comment by kurnevsky** [comment]
Ugh, nix-review builds all changed packages, which is a lot. I tested it with such change in my configs: https://github.com/kurnevsky/nixfiles/commit/3e8bfd62c98af1ecb86075c045bb1267c874de60

**Comment by kurnevsky** [comment]
Also there is already a [2.4.3 version](https://lists.gnupg.org/pipermail/gnupg-announce/2023q3/000480.html) but it doesn't include this patch.

**Comment by nagy** [comment]
Emacs only workaround:

```elisp
(fset 'epg-wait-for-status 'ignore)
```

If this only affects Emacs, should we not patch Emacs itself, rather than patching this security tool for everybody else?

**Comment by kurnevsky** [comment]
This patch is already merged to the gnupg master, it's just not backported to the STABLE-BRANCH-2-4 branch for some reason.

Have you tested this emacs-only wokaround? Could it break something if we completely remove waiting for gpg statuses?

**Comment by kurnevsky** [comment]
Looks like this removing of waiting might break some interaction with pinentry. At least there is a note with suggested patch for emacs there: `Waiting for BEGIN_ENCRYPTION is not needed when it's for public-key encryption, because there is no change for pinentry to ask passphrase.`

**Comment by nagy** [comment]
> This patch is already merged to the gnupg master, it's just not backported to the STABLE-BRANCH-2-4 branch for some reason. They also suggest to apply it for distributions in one of the latest comments in that thread. 

Oh ok, it that case this approach is fine. Should we leave a note, that this patch can be removed with a future version?

> Have you tested this emacs-only wokaround?

Yes it works fine for me.

> Could it break something if we completely remove waiting for gpg statuses?

Of course, but lets not use that approach then.

**Comment by kurnevsky** [comment]
> Should we leave a note, that this patch can be removed with a future version?

Added such note.

File: pkgs/tools/security/gnupg/24.nix [hunk 1]
```diff
@@ -35,6 +35,12 @@ stdenv.mkDerivation rec {
     ./24-allow-import-of-previously-known-keys-even-without-UI.patch
     # Patch for DoS vuln from https://seclists.org/oss-sec/2022/q3/27
     ./v3-0001-Disallow-compressed-signatures-and-certificates.patch
+    # Patch for emacs easypg https://dev.gnupg.org/T6481
+    # It's merged to gnupg master so can be removed in the next major version
+    (fetchpatch {
+      url = "https://files.gnupg.net/file/data/iapdg4omgofpqs5c6gd5/PHID-FILE-pym22ajak2knrci6sc2h/file";
```
Comment (nagy) [inline pkgs/tools/security/gnupg/24.nix:41]: Could we use this url rather `https://dev.gnupg.org/rG2f872fa68c6576724b9dabee9fb0844266f55d0d?diff=1` ? This should redirect to the same content , so the sha256 sum can stay the same. But this url allows to trace this back to a commit ( by removing the `?diff=1`) and it does not look like some random attachment that somebody might have posted to a bugtracker.

**Review by nagy — CHANGES_REQUESTED** [review]

File: pkgs/tools/security/gnupg/24.nix (hunk 1 above)
Comment (kurnevsky) [reply pkgs/tools/security/gnupg/24.nix]: Sure.

**Review by nagy — APPROVED** [review]

**Comment by dschrempf** [comment]
I think this can and should be merged.

How do we ensure that the patch is going to be removed after the next gnupg major release? Will the Nixpkgs gnupg build fail, because the patch fails; and so, we have to intervene manually and remove the patch?

**Comment by nagy** [comment]
Can we try something like :
```nix
  ++ lib.optionals (lib.versionOlder version "3.0.0") [
    (fetchpatch {
      ...
    })
```
?

This is with my current observation, that this commit has not been merged into the 2.4 stable branch. So it will only be seen in 3.0 . This way it would not break and therefore, we would not be notified, but this allows us eventually to remove the patch without having to go through another staging iteration again.

**Comment by kurnevsky** [comment]
Shouldn't it fail to apply when the upstream version gets it? So then whoever willb e updating it will see that it has to be removed.

**Comment by nagy** [comment]
It may also be applied by a bot, which then just fails on it's own without opening a PR and people have to investigate. But I dont have a preference either way. But I think we need to ask other people in case of such a core package.

**Comment by SuperSandro2000** [comment]
> This is with my current observation, that this commit has not been merged into the 2.4 stable branch. So it will only be seen in 3.0 . This way it would not break and therefore, we would not be notified, but this allows us eventually to remove the patch without having to go through another staging iteration again.

We don't have both versions intree and we want to notice when an upgrade lands that makes the patch redundant. 

The bigger problem right now is: why does the manual not build?

**Comment by dschrempf** [comment]
Good question. The logs are not very conclusive. During the tests, building `aws-cdk-cpp` fails, but the error is not in the logs (it could be retreived by `'nix log /nix/store/2chagxkzpf7bc0d25mjng3sn6gs0bj9z-aws-sdk-cpp-1.11.118.drv').

File: pkgs/tools/security/gnupg/24.nix [hunk 2]
```diff
@@ -35,6 +35,12 @@ stdenv.mkDerivation rec {
     ./24-allow-import-of-previously-known-keys-even-without-UI.patch
     # Patch for DoS vuln from https://seclists.org/oss-sec/2022/q3/27
     ./v3-0001-Disallow-compressed-signatures-and-certificates.patch
+    # Patch for emacs easypg https://dev.gnupg.org/T6481
+    # It's merged to gnupg master so can be removed in the next major version
+    (fetchpatch {
+      url = "https://dev.gnupg.org/rG2f872fa68c6576724b9dabee9fb0844266f55d0d?diff=1";
+      sha256 = "sha256-irxxBBO0MqHaUj0a0qZ8yfhKp4wXRdQErj3eD/ZF66I=";
+    })
```
Comment (vcunat) [inline pkgs/tools/security/gnupg/24.nix:43]: This URL generates patch in a bit unusual way.  The whole file is in there (264K), not just parts around the changes.

**Comment by dschrempf** [comment]
Reason for closure: Fixed in GnuPG 2.4.4. https://dev.gnupg.org/T6481

**Pull request closed without merging**